document config upgrader (#288)
cylc/cylc-doc: src/7-to-8/major-changes/before-you-start.rst

.. _AutoConfigUpgrades:

Cylc Configuration Upgrader
===========================

.. admonition:: Does This Change Affect Me?
   :class: tip

   This change may affect you:

   - If you are working with workflows written before Cylc 7.
   - If you hope to use Cylc 9.

   You have been affected by this change if you see a message such as:

   .. code-block::

      > cylc validate my_workflow
      IllegalItemError: reallyoldconfig

Overview
--------

> Cylc has a built in configuration upgrader. Cylc can upgrade Cylc 7

Author: ```suggestion
Cylc has a built in configuration upgrader. Cylc 8 can upgrade Cylc 7
```

> Cylc has a built in configuration upgrader. Cylc can upgrade Cylc 7
> workflows to Cylc 8 workflows. Cylc cannot upgrade Cylc 6 or earlier

Reviewer: This might give the impression it will upgrade the file itself?
```suggestion
Cylc has a built in configuration upgrader. Cylc 8 can upgrade Cylc 7
workflows at runtime. Cylc cannot upgrade Cylc 6 or earlier
```

> workflows to Cylc 8 workflows. Cylc cannot upgrade Cylc 6 or earlier

Author: ```suggestion
workflows to Cylc 8 workflows. Cylc 8 cannot upgrade Cylc 6 or earlier
```
```

workflows to Cylc 8.

Solution
--------

To avoid problems with old config items you should validate your workflow using

> Cylc 7. Look for deprecation warnings and change your configuration to avoid
> these warnings.

Reviewer: ```suggestion
Cylc 7. Look for deprecation warnings and edit the workflow configuration to
eliminate these warnings.
```

Example
-------

Consider this configuration:

.. code-block:: cylc

   [scheduling]
       initial cycle point = 11000101T00

   [[dependencies]]
       [[[R1]]]
           graph = task

   [runtime]
       [[task]]
           pre-command scripting = echo "Hello World"

Running ``cylc validate`` on this configuration at Cylc 7 we see that the
workflow is valid, but we are warned that ``pre-command scripting``

> was replaced by ``pre-script`` at 6.4.0:

Reviewer 2: When I recreated this example I get:
```WARNING - deprecated graph items were automatically upgraded in "workflow definition":
	 * (8.0.0) [scheduling][dependencies][X]graph -> [scheduling][graph]X - for X in:
	       R1
IllegalItemError: [runtime][task]pre-command scripting
```

Strictly speaking it doesn't warn about the replacement being `pre-script`, and no mention of the 6.4.0 (I am on master on cylc-flow).
I wonder if it should?

Reviewer: The instructions say validate using Cylc 7 first, perhaps it should be in bold if it's too easily missed

Author: Deliberately glossing over the warning about the graphing: It's only warning at Cylc 8, not an error. 
Boldening the version seems wise.

Author: Had another look. Sorry fiddled with it a fair bit.
